rfcs/rfc-007-dynamic-program-dispatch.md

```diff
@@ -0,0 +1,183 @@
+
+This document highlights a set of changes to enable a framework for Solana
+programs to be composed and dynamically dispatched.
+
+### Account additions
+
+```diff
+diff --git a/common/src/account.rs b/common/src/account.rs
+index 5abd642..89290ea 100644
+--- a/common/src/account.rs
++++ b/common/src/account.rs
+@@ -11,6 +11,12 @@ pub struct Account {
+     pub userdata: Vec<u8>,
+     /// contract id this contract belongs to
+     pub program_id: Pubkey,
++
++    /// this account contains a program (and is strictly read-only)
++    pub executable: bool,
++
++    /// the loader for this program
++    pub loader_program_id: Option<Pubkey>,
+ }
+
+ impl Account {
+@@ -19,6 +25,8 @@ impl Account {
+             tokens,
+             userdata: vec![0u8; space],
+             program_id,
++            false,
++            None,
+         }
+     }
+ }
+```
+
+
+The executable flag may only be set by the `program_id` the account belongs to.
+Once the executable flag is set:
+1. The bank enforces the account userdata read-only even by `program_id`
+2. The executable flag cannot be unset
+3. The account may be used as input to `SystemProgram::Spawn` to forge a new `program_id`
+
+The `loader_program_id` field is set by `SystemProgram::Spawn` and is used by
+bank to locate the required loader during transaction dispatch.
+
+### Common Loader Interface
+
+```rust
+//! Instructions that must be implemented by all Loaders
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub enum LoaderInstruction {
+    /// Write program data into an account
+    ///
+    /// * key[0] - the account to write into.
+    ///
+    /// The transaction must be signed by key[0]
+    Write { offset: u32, prog: Vec<u8> },
+
+    /// Prepare an account loaded with program data for execution.
+    /// The exact preparation steps is loader specific but on success the loader must set the executable
+    /// bit of the Account
+    ///
+    /// * key[0] - the account to prepare for execution
+    ///
+    /// The transaction must be signed by key[0]
+    Prepare,
```

Comment:
"Prepare" seems like the first thing one would need to do. How about a term like "Finish" or "MakeExecutable"?

Reply:
Yeah, I agree.  `Finalize`?.